### DISCUSSION TITLE
make shutdown more robust

### DISCUSSION THREAD

**Discussion by WolfgangFahl** [Kludex/uvicorn#1617] [Potential Issue]
File "/Users/wf/Library/Python/3.10/lib/python/site-packages/uvicorn/server.py", line 256, in shutdown
    for server in self.servers:
AttributeError: 'Server' object has no attribute 'servers'

the code might want to check that self.servers is set before trying the loop. A warning that there are now servers would IMHO be better than an Exception.

**Comment by Kludex** [comment]
How can I reproduce it?

**Reply by JimiC** [reply]
This happens when you use programmatically `Multiprocess` to spawn several instances of the `Server` and programmatically call `shutdown` method.

Here's a test that produces the error.
```python
import asyncio

from uvicorn import Config, Server
from uvicorn.supervisors import Multiprocess


async def test():
    server_config = Config(app="fastapi:FastAPI", factory=True, workers=2)
    server = Server(config=server_config)
    sock = server.config.bind_socket()
    Multiprocess(server.config, target=server.run, sockets=[sock]).run()
    await server.shutdown()


if __name__ == "__main__":
    asyncio.run(test())
```

After some more digging I found that the cause is exactly the fact that we are calling `server.shutdown` and that `server` (we create) has no `servers` nor `lifespan` attributes.

**Comment by WolfgangFahl** [comment]
https://github.com/justpy-org/justpy/issues/474 uses a workaround for #1103 since the psutil version does not work on MacOS. So if #1103 would be fixed the workaround would not be necessary. The workaround does not work properly and doesn't call shutdown properly since I didn't understand what do do given as the lenghty discussion of #1103 didn't come to a complexity hiding solution yet. So I end up trying my best and get multiple error mesages one of which is the one mentioned in this discussion. It all boils down to a proper shut down being possible programmatically in a hassle free and platform independent way.
